docs/modules/candid-guide/index.adoc

```diff
@@ -0,0 +1,1098 @@
++
+ * The result type list may be shortened
+ * The result type list may be extended with optional arguments (type `opt …`)
+ * The parameter type list may be extended.
+ * Existing parameter types may be change to to a subtype.
+ * Existing result types may be changed to a supertype(!).
+
+Corresponding Motoko type::
+
+Candid function types correspond to `shared` Motoko functions, with the result type wrapped in `async`. Arguments resp. results become tuples, unless there is exactly one, in which case it is used directly:
++
+....
+type F0 = func () -> ();
+type F1 = func (text) -> (text);
+type F2 = func (text, bool) -> () oneway;
+type F3 = func (text) -> () oneway;
+type F4 = func () -> (text) query;
+....
+corresponds in Motoko to
+....
+type F0 = shared () -> async ();
+type F1 = shared Text -> async Text;
+type F2 = shared (Text, Bool) -> ();
+type F3 = shared query () -> async Text;
+....
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-service]
+=== Type service {…}
+
+Services may want to pass around references to not just individual functions (using the <<type-func,`func` type>>), but references to whole services. In this case, Candid types can be used to declare the complete interface of such a service.
+
+Type syntax::
++
+....
+service {
+  add : (nat) -> ();
+  subtract : (nat) -> ();
+  get : () -> (int) query;
+  subscribe : (func (int) -> ()) -> ();
+}
+....
++
+See Section “<<candid-service-descriptions>>” for more details on the syntax of a service type.
+
+Textual syntax::
++
+....
+service "w7x7r-cok77-xa"
+service "zwigo-aiaaa-aaaaa-qaa3a-cai"
+service "aaaaa-aa"
+....
+
+Subtypes::
+
+The subtypes of a service type are those service types that possibly have additional method, and where the type of existing methods is changed to a subtype.
++
+This is exactly the notion that shows up in Section “<<upgrades,safe upgrades>>”.
+
+Supertypes::
+
+The supertypes of a service type are those service types that may have some methods removed, and the type of existing methods are changed to a supertype.
+
+Corresponding Motoko type::
+
+Service types in Candid correspond directly to `actor` types in Motoko:
++
+....
+actor {
+  add : shared Nat -> async ()
+  subtract : shared Nat -> async ();
+  get : shared query () -> async Int;
+  subscribe : shared (shared Int -> async ()) -> async ();
+}
+....
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-principal]
+=== Type principal
+
+The Internet computer uses _principals_ as the common scheme to idnetify canisters, users and other entities.
+
+Type syntax::
+
+`principal`
+
+
+Textual syntax::
++
+....
+principal "w7x7r-cok77-xa"
+principal "zwigo-aiaaa-aaaaa-qaa3a-cai"
+principal "aaaaa-aa"
+....
+
+
+Corresponding Motoko type::
+
+`Principal`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-null]
+=== Type null
```

Comment:
Maybe move this further up (e.g., next to opt?), since it's referenced in various previous sections.